### PR TITLE
Add Sum Overload Accepting Function

### PR DESCRIPTION
**Pull request by PieceMaker**
This set of changes adds the overload requested in #26.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -35,15 +35,22 @@ interface Array<T> {
...
+    keySelector: ((x: T) => K),
+    comparer: ((x: K, y: K) => number)
+  ): Array<T>;
   Prepend<T>(element: T): Array<T>;
   Reverse<T>(): Array<T>;
   Select<T, U>(xform: (x: T) => U): Array<U>;
   Take<T>(count: number): Array<T>;
   TakeLast<T>(count: number): Array<T>;
   Sum(): number;
+  Sum(predicate: (x: T) => number): number;
```
Comment (jreina) [inline index.d.ts]:
```suggestion
  Sum(selector: (x: T) => number): number;
```

File: src/Sum.js [hunk 2]
```diff
@@ -1,13 +1,35 @@
 /**
- * @template number?
+ * @template T
+ * @param {Array<T>} array
+ * @param {((x: T) => number)} predicate
```
Comment (jreina) [inline src/Sum.js]:
```suggestion
 * @param {((x: T) => number)} selector
```

File: src/Sum.js [hunk 3]
```diff
@@ -1,13 +1,35 @@
...
+ * @returns {number}
+ */
+function _predicateSum(array, predicate) {
```
Comment (jreina) [inline src/Sum.js]:
```suggestion
function _predicateSum(array, selector) {
```

File: src/Sum.js [hunk 4]
```diff
@@ -1,13 +1,35 @@
...
+  let predicateResult = array.map(predicate);
```
Comment (jreina) [inline src/Sum.js]:
```suggestion
  const numbers = array.map(selector);
```
```suggestion
  const xs = array.map(selector);
```

File: src/Sum.js [hunk 5]
```diff
@@ -1,13 +1,35 @@
...
+  return predicateResult.Sum();
+}
+/**
+ * @template T
+ * @param {((x: T) => number)} predicate
```
Comment (jreina) [inline src/Sum.js]:
```suggestion
 * @param {((x: T) => number)} selector
```

**Review by jreina — CHANGES_REQUESTED** [review]
`predicate` should be `selector` in all cases. You can probably find and replace this but I wanted to try the suggested changes feature.

**Comment by PieceMaker** [comment]
Changes made.

**Pull request merged**